machineconfiguration/v1/types_machineosconfig.go

```diff
@@ -0,0 +1,247 @@
+	// This is used as a label in the Containerfile that builds the OS image.
+	// +kubebuilder:validation:MaxLength:=253
+	// +kubebuilder:validation:XValidation:rule="!format.dns1123Subdomain().validate(self).hasValue()",message="a lowercase RFC 1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character."
+	// +optional
+	ReleaseVersion string `json:"releaseVersion,omitempty"`
+	// containerFile describes the custom data the user has specified to build into the image.
+	// This is also commonly called a Dockerfile and you can treat it as such. The content is the content of your Dockerfile.
+	// See https://github.com/containers/common/blob/main/docs/Containerfile.5.md for the spec reference.
+	// you can specify up to 7 containerFiles
+	// +patchMergeKey=containerfileArch
+	// +patchStrategy=merge
+	// +listType=map
+	// +listMapKey=containerfileArch
+	// +kubebuilder:validation:MinItems=0
+	// +kubebuilder:validation:MaxItems=7
+	// +optional
+	Containerfile []MachineOSContainerfile `json:"containerFile" patchStrategy:"merge" patchMergeKey:"containerfileArch"`
+}
+
+// BuildOutputs holds all information needed to handle booting the image after a build
+type BuildOutputs struct {
+	// currentImagePullSecret is the secret used to pull the final produced image.
+	// Must live in the openshift-machine-config-operator namespace,
+	// the final image push and pull secrets should be separate for security concerns. If the final image push secret is somehow exfiltrated,
+	// that gives someone the power to push images to the image repository. By comparison, if the final image pull secret gets exfiltrated,
+	// that only gives someone to pull images from the image repository. It's basically the principle of least permissions.
+	// This pull secret will be used on all nodes in the pool. These nodes will need to pull the final OS image and boot into it using rpm-ostree or bootc.
+	// +optional
+	CurrentImagePullSecret *ImageSecretObjectReference `json:"currentImagePullSecret,omitempty"`
+}
+
+type MachineOSImageBuilder struct {
+	// imageBuilderType specifies the backend to be used to build the image.
+	// +kubebuilder:validation:Enum:=JobImageBuilder
+	// Valid options are: JobImageBuilder
+	// +required
+	ImageBuilderType MachineOSImageBuilderType `json:"imageBuilderType"`
+}
+
+// MachineOSContainerfile contains all custom content the user wants built into the image
+type MachineOSContainerfile struct {
+	// containerfileArch describes the architecture this containerfile is to be built for.
+	// This arch is optional. If the user does not specify an architecture, it is assumed
+	// that the content can be applied to all architectures, or in a single arch cluster: the only architecture.
+	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;AArch64;x86_64;NoArch
```

Comment:
Why do we have these duplicates? That means an end user can configure the same meaning, in multiple ways, that's confusing, we should avoid that
```suggestion
	// +kubebuilder:validation:Enum:=ARM64;AMD64;PPC64LE;S390X;NoArch
```